custom_components/isy994/services.py

```diff
@@ -0,0 +1,260 @@
+        SERVICE_SEND_COMMAND, SERVICE_SEND_COMMAND_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_BEEP, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_BRIGHTEN, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_DEVICE_QUERY, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_DIM, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_DISABLE, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_ENABLE, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_FADE_DOWN, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_FADE_STOP, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_FADE_UP, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_FAST_OFF, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+    platform.async_register_entity_service(
+        SERVICE_FAST_ON, SERVICE_SIMPLE_SCHEMA, device_service_handler
+    )
+
+
+async def async_setup_light_services(hass: HomeAssistantType):
+    """Create device-specific services for the ISY Integration."""
+    platform = entity_platform.current_platform.get()
+
+    async def light_service_handler(entity, service_call):
+        """Handle device-specific service calls."""
+        # pylint: disable=protected-access
+        if not hasattr(entity, "_node") or not hasattr(
+            entity._node, service_call.service
+        ):
+            _LOGGER.warning(
+                "Invalid Service Call %s for device %s.",
+                service_call.service,
+                entity.entity_id,
+            )
+            return
+
+        if service_call.service in [SERVICE_SET_RAMP_RATE, SERVICE_SET_ON_LEVEL]:
+            value = service_call.data.get(ATTR_VALUE)
+            _LOGGER.debug(
+                "Sending command %s to device %s with value %s.",
+                service_call.service,
+                entity.entity_id,
+                value,
+            )
+            await hass.async_add_executor_job(
+                ft.partial(getattr(entity._node, service_call.service), value)
+            )
+            return
+
+        _LOGGER.debug(
+            "Sending command %s to device %s.", service_call.service, entity.entity_id
+        )
+        await hass.async_add_executor_job(getattr(entity._node, service_call.service))
+
+    platform.async_register_entity_service(
+        SERVICE_SET_ON_LEVEL, SERVICE_SET_VALUE_SCHEMA, light_service_handler
```

Comment:
```suggestion
        SERVICE_SET_ON_LEVEL, {vol.Required(ATTR_VALUE): vol.All(vol.Coerce(int), vol.Range(0, 255))}, "set_on_level"
```

If you change this to a string it will call `def set_on_level(...)` directly on the entity with the service inputs

Reply:
...and add this to the `ISYLightDevice`

```
    def set_on_level(self, **kwargs) -> None:
        """Set on level of the ISY994 light device."""
     ...
```

Reply:
I saw that method but was trying to avoid adding all of the functions to the LightDevice and ISYDevice classes since they could be called through the _node, but I guess it's probably better to just add them and avoid using `getattr` to call them.

Reply:
OK. I updated the 4 Light-based services to just call directly on the LightDevice functions.

I haven't changed the more generic device functions yet. I'm tempted to leave these as is because it is a consolidated check to make sure the underlying PyISY class has the function before calling it. If I added each function to ISYDevice, I would have to add the attribute check to each individual function since Variables and Programs won't have these functions in PyISY (and everything is inherited from ISYDevice).

Thoughts?

Reply:
Seems fine to me.

Another option which seems a bit cleaner would be to make a new base class maybe `ISYServiceDevice` that has `ISYDevice` as a base class and only switch the ones you want to it.

Reply:
I'll stew on it for a day. There's definitely room for cleanup on the inheritance--I think there needs to be a common base for ISYProgram and ISYVariable to clean up the code--I may get to that tomorrow. 

But as far as the services go: I did find that the core call doesn't guard against the function not existing in the entity; it just calls getattr() and an error will get thrown if it's not there, which I guess is fine... since you shouldn't be calling the service on that entity anyways. I was just trying to log it a little more gracefully.

https://github.com/home-assistant/core/blob/f4f2aff5b69c5be44f6148964ccded2e600cd104/homeassistant/helpers/service.py#L478